src/calliope/model.py

```diff
@@ -567,3 +574,181 @@ def _recalculate_storage_initial(self, results: xr.Dataset) -> xr.DataArray:
+        LOGGER.info(
+            f"Optimisation model | Running SPORES with `{spores_config.scoring_algorithm}` scoring algorithm."
+        )
+        for spore in spore_range:
+            LOGGER.info(f"Optimisation model | Running SPORE {spore}.")
+            self._spores_update_model(baseline_results, results_list, spores_config)
+
+            iteration_results = self.backend._solve(solver_config, warmstart=False)
+            results_list.append(iteration_results)
+
+            if spores_config.save_per_spore_path is not None:
+                LOGGER.info(f"Optimisation model | Saving SPORE {spore} to file.")
+                iteration_results.assign_coords(spores=spore).to_netcdf(
+                    spores_config.save_per_spore_path / f"spore_{spore}.nc"
+                )
+
+        spores_dim = pd.Index(["baseline", *spore_range], name="spores")
+        results = xr.concat(results_list, dim=spores_dim, combine_attrs="no_conflicts")
+        results.attrs["termination_condition"] = ",".join(
+            set(result.attrs["termination_condition"] for result in results_list)
+        )
+
+        return results
+
+    def _spores_update_model(
+        self,
+        baseline_results: xr.Dataset,
+        all_previous_results: list[xr.Dataset],
+        spores_config: config_schema.SolveSpores,
+    ):
+        """Assign SPORES scores for the next iteration of the model run.
+
+        Algorithms applied are based on those introduced in <https://doi.org/10.1016/j.apenergy.2023.121002>.
+
+        Args:
+            baseline_results (xr.Dataset): The initial results (before applying SPORES scoring)
+            all_previous_results (list[xr.Dataset]):
+                A list of all previous iterations.
+                 This includes the baseline results, which will be the first item in the list.
+            spores_config (config_schema.SolveSpores):
+                The SPORES configuration.
+        """
+
+        def _score_integer() -> xr.DataArray:
+            """Integer scoring algorithm."""
+            previous_cap = latest_results["flow_cap"].where(spores_techs)
+
+            # Make sure that penalties are applied only to non-negligible deployments of capacity
+            min_relevant_size = spores_config.score_threshold_factor * previous_cap.max(
+                ["nodes", "techs"]
+            )
+
+            new_score = (
+                # Where capacity was deployed more than the minimal relevant size, assign an integer penalty (score)
+                previous_cap.where(previous_cap > min_relevant_size)
+                .clip(min=1, max=1)
+                .fillna(0)
+                .where(spores_techs)
+            )
+            return new_score
+
+        def _score_relative_deployment() -> xr.DataArray:
+            """Relative deployment scoring algorithm."""
+            previous_cap = latest_results["flow_cap"].where(spores_techs)
+            relative_cap = previous_cap / self.inputs["flow_cap_max"].where(
```

Comment:
A problem that I've never addressed properly is what happens if no cap_max is defined (i.e., it is `inf` as per the default). Normally, I'd just make sure that every tech has a cap_max, even when this is not due to obvious reasons (e.g., land constraints). For instance, batteries are typically unconstrained in our models, and I would arbitrarily assign them a cap_max that makes sense based on the order of magnitude of other things. 

To automate this, should we say that where no cap_max exists, the maximum relative potential is capped at a value that has the same order of magnitude as the highest cap_max in the model?